packages/core/src/shared/ui/nestedWizardPrompter.ts

```diff
@@ -0,0 +1,63 @@
+/*!
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { Wizard, WizardOptions } from '../wizards/wizard'
+import { Prompter } from './prompter'
+import { WizardPrompter } from './wizardPrompter'
+import { createHash } from 'crypto'
+
+/**
+ * An abstract class that extends the base Wizard class plus the ability to
+ * use other wizard classes as prompters
+ */
+export abstract class NestedWizard<T> extends Wizard<T> {
+    /**
+     * Map to store memoized wizard instances using SHA-256 hashed keys
+     */
+    private wizardInstances: Map<string, any> = new Map()
+
+    public constructor(options?: WizardOptions<T>) {
+        super(options)
+    }
+
+    /**
+     * Creates a prompter for a wizard instance with memoization.
+     *
+     * @template TWizard - The type of wizard, must extend Wizard<TState>
+     * @template TState - The type of state managed by the wizard
+     *
+     * @param wizardClass - The wizard class constructor
+     * @param args - Constructor arguments for the wizard instance
+     *
+     * @returns A wizard prompter to be used as prompter
+     *
+     * @example
+     * // Create a prompter for SyncWizard
+     * const prompter = this.createWizardPrompter<SyncWizard, SyncParams>(
+     *     SyncWizard,
+     *     template.uri,
+     *     syncUrl
+     * )
+     *
+     * @remarks
+     * - Instances are memoized using a SHA-256 hash of the wizard class name and arguments
+     * - The same wizard instance is reused for identical constructor parameters for restoring wizard prompter
+     *   states during back button click event
+     */
+    protected createWizardPrompter<TWizard extends Wizard<TState>, TState>(
+        wizardClass: new (...args: any[]) => TWizard,
+        ...args: ConstructorParameters<new (...args: any[]) => TWizard>
+    ): Prompter<TState> {
+        const memoizeKey = createHash('sha256')
+            .update(wizardClass.name + JSON.stringify(args))
+            .digest('hex')
+
+        if (!this.wizardInstances.get(memoizeKey)) {
+            this.wizardInstances.set(memoizeKey, new wizardClass(...args))
```

Comment:
since wizardClass already gets the args, would it make sense for it to define a `getId()` or `getHash()` that provides its hash? Instead of managing that here.

It seems that javascript doesn't have Java's `hashCode()` and `equals()` conventions, but we can mimic it I guess.

Reply:
Do you suggest moving the hashing logic from `NestedWizard` and move that to Wizard class as static function? 
I say static because we might want to check hash before instantiating an object. 

But this would not be that much different from having the hashing logic here in `NestedWizard`

Reply:
> say static because we might want to check hash before instantiating an object.

Oh, I didn't think about that. But it still seems to make sense for the Wizard to define its own hash. After all, this may be used for other purposes, but currently it's private to this one consumer.

Reply:
Remove this class after moving logic here to `Wizard` class.